### PR TITLE
Add OSL vector transforms. Fix OSL NDC & raster transforms.

### PR DESCRIPTION
**Pull request by est77**

### CODE REVIEW

File: src/appleseed/renderer/kernel/rendering/rendererservices.cpp [hunk 1]
```diff
@@ -343,36 +367,57 @@ bool RendererServices::transform_points(
...
+            {
+                m_camera->project_point(
+                    time,
+                    Vector3d(sg->P.x, sg->P.y, sg->P.z),
+                    ndc_p);
+            }
+
+            for (size_t i = 0; i < npoints; ++i)
+            {
+                OSL::Vec3 Q = sg->P + Pout[i];
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/rendererservices.cpp]: Make const.

File: src/appleseed/renderer/kernel/rendering/rendererservices.cpp [hunk 2]
```diff
@@ -307,6 +307,31 @@ bool RendererServices::get_inverse_matrix(
...
+    // We only transform points and vectors for now.
+    return
+        vectype == OSL::TypeDesc::POINT ||
+        vectype == OSL::TypeDesc::VECTOR;
+}
+
+void transform(
+    const OSL::Vec3&            Pin,
+    const OSL::Matrix44&        m,
+    OSL::TypeDesc::VECSEMANTICS vectype,
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/rendererservices.cpp]: Make const.

File: src/appleseed/renderer/kernel/rendering/rendererservices.cpp [hunk 3]
```diff
@@ -307,6 +307,31 @@ bool RendererServices::get_inverse_matrix(
     return OSL::RendererServices::get_inverse_matrix(sg, result, to);
 }
 
+namespace
+{
+
+bool can_transform(OSL::TypeDesc::VECSEMANTICS vectype)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/rendererservices.cpp]: Make argument const.

File: src/appleseed/renderer/kernel/rendering/rendererservices.cpp [hunk 4]
```diff
@@ -343,36 +367,57 @@ bool RendererServices::transform_points(
...
+
+                Vector2d ndc_q(-1.0, -1.0);
+                m_camera->project_point(time, Vector3d(Q.x, Q.y, Q.z), ndc_q);
+                Pout[i] = OSL::Vec3(
+                    static_cast<float>(ndc_q[0] - ndc_p[0]),
+                    static_cast<float>(ndc_q[1] - ndc_p[1]),
+                    0.0f);
+            }
+        }
+        else // if (vectype == OSL::TypeDesc::POINT)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/rendererservices.cpp]: Replace comment by assertion.

File: src/appleseed/renderer/kernel/rendering/rendererservices.cpp [hunk 5]
```diff
@@ -343,36 +367,57 @@ bool RendererServices::transform_points(
...
                     0.0f);
             }
-            else
-                Pout[i] = OSL::Vec3(-1.0f, -1.0f, 0.0f);
         }
 
         if (to == g_raster_ustr)
         {
             // Transform from NDC to raster.
             for (int i = 0; i < npoints; ++i)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/rendererservices.cpp]: What's the type of `npoints`? I haven't checked but maybe `i` needs to be of type `size_t`.

**Comment by est77** [comment]
Fixed and rebased

**Comment by luisbarrancos** [comment]
@est77 thanks for the swift (no pun intended) implementation, it'll be handy for the projection and matte painting tools.

**Pull request merged**